### PR TITLE
Ai implementation continue

### PR DESCRIPTION
**Pull request by C-gyorfi**
Partially implemented AI, let's say it's an artificial unintelligence
Find winning combinations use case implemented
Done some refactoring, but we are planning to do more :)

### CODE REVIEW

**Comment by chubberlisk** [comment]
This [`README.md`](https://github.com/C-gyorfi/tic-tac-toe/blob/c8bf04e6ad5a0789902ea112cc715284e91b1df8/README.md) looks somewhat familiar? ;P

**Comment by chubberlisk** [comment]
https://github.com/C-gyorfi/tic-tac-toe/blob/c8bf04e6ad5a0789902ea112cc715284e91b1df8/lib/use_case/find_winning_combinations.rb#L16-L22

I'm not sure that the above snippet of code is very Ruby-like. I think the same logic can be achieved without breaking a loop and incrementing a counter variable. A possible (and tested) alternative below:

```ruby
(0...board.length).step(@width) do |index|
  winning_combinations << (index...index + @width).to_a
end
```

Check out https://ruby-doc.org/core-2.2.0/Range.html#method-i-step to find out more about the `Range#step` method in Ruby.

**Comment by chubberlisk** [comment]
Similarly, in the `#vertical_winnings_for` method.

https://github.com/C-gyorfi/tic-tac-toe/blob/c8bf04e6ad5a0789902ea112cc715284e91b1df8/lib/use_case/find_winning_combinations.rb#L31-L36

You could do:

```ruby
(column...board.length).step(@width) do |index|
  current_comumn << index
  winning_combinations << current_comumn
end
```

File: lib/use_case/ai_response_board.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  end
+
+  private
+
+  def block_opponent(board)
+    @chance_to_block = []
+    @winning_combinations.each do |combination|
+      next if already_blocked(combination, board)
+
+      @chance_to_block << combination.select { |i| i if board[i] != 'X' }
```
Comment (chubberlisk) [inline lib/use_case/ai_response_board.rb]:
```suggestion
      @chance_to_block << combination.reject { |i| board[i] == 'X' }
```

**Comment by chubberlisk** [comment]
https://github.com/C-gyorfi/tic-tac-toe/blob/c8bf04e6ad5a0789902ea112cc715284e91b1df8/lib/use_case/ai_response_board.rb#L27-L29

Alternatively:

```ruby
@chance_to_block.select { |chances| chances.length == 1 }.first.first
```

**Comment by C-gyorfi** [comment]
> This [`README.md`](https://github.com/C-gyorfi/tic-tac-toe/blob/c8bf04e6ad5a0789902ea112cc715284e91b1df8/README.md) looks somewhat familiar? ;P

**Comment by C-gyorfi** [comment]
> This [`README.md`](https://github.com/C-gyorfi/tic-tac-toe/blob/c8bf04e6ad5a0789902ea112cc715284e91b1df8/README.md) looks somewhat familiar? ;P

**Comment by C-gyorfi** [comment]
> [tic-tac-toe/lib/use_case/find_winning_combinations.rb](https://github.com/C-gyorfi/tic-tac-toe/blob/c8bf04e6ad5a0789902ea112cc715284e91b1df8/lib/use_case/find_winning_combinations.rb#L16-L22)
> 
> Lines 16 to 22 in [c8bf04e](/C-gyorfi/tic-tac-toe/commit/c8bf04e6ad5a0789902ea112cc715284e91b1df8)
> 
>  i = 0 
>  loop do 
>    break if i >= board.length 
>   
>    winning_combinations << (i...i + @width).to_a 
>    i += @width 
>  end 
> I'm not sure that the above snippet of code is very Ruby-like. I think the same logic can be achieved without breaking a loop and incrementing a counter variable. A possible (and tested) alternative below:
> 
> ```ruby
> (0...board.length).step(@width) do |index|
>   winning_combinations << (index...index + @width).to_a
> end
> ```
> 
> Check out https://ruby-doc.org/core-2.2.0/Range.html#method-i-step to find out more about the `Range#step` method in Ruby.

I knew there must be a more elegant way to do this

File: lib/use_case/ai_response_board.rb (hunk 1 above)
Comment (C-gyorfi) [reply lib/use_case/ai_response_board.rb]: Good point! I think I should learn more about loops in Ruby. I never heard about .step before and .reject reads much better in this case.

**Pull request merged**